#-------------------------------------#
#              * scDRS *              #
#         Gene set enrichment         #
#-------------------------------------#
# 1 create gene set (config generated by gwas summary)
magma=/public/home/shilulu/software/MAGMA_v1.10/magma
eur="/public/share/wchirdzhq2022/Wulab_share/1000GenomePhase3_Ref_hg37/g1000_eur/g1000_eur"
geneloc="/public/home/shilulu/software/MAGMA_v1.10/NCBI37.3.SYMB.gene.loc"
prefix=$(basename "$eur")

#- gene annotation
qsubshcom "$magma --annotate window=35,10 \
--snp-loc $eur.bim \
--gene-loc $geneloc \
--out $prefix" 1 100G magma 1:00:00 ""

#- SNP p extract
gwas="/public/home/shilulu/Wulab/sll/ARHL/NC_sup_test/01.meta/All_MVP_Trpchevska_De-Angelis_BBJ_filter.txt"
trait="ARHL"
awk 'NR > 1 {print $1"\t"$7"\t"$8}' $gwas > $trait.pval

#- gene-based association study
qsubshcom "$magma --bfile $eur \
--pval $trait.pval ncol=3 \
--gene-annot $prefix.genes.annot
--out $trait" 1 100G magma 1:00:00 ""


# extract GENE and ZSCORE of trait
awk 'NR==1 {print "GENE\tARHL"} NR>1 {print $1"\t"$8}' $trait.genes.out > $trait.geneset.zscore

# 2 munge gene sets
conda activate scdrs

qsubshcom "scdrs munge-gs \
    --out-file ${trait}.scdrs \
    --zscore-file ${trait}.geneset.zscore \
    --weight zscore \
    --n-max 1000" 1 10G scdrs_munge 1:00:00 ""

# 3. run scDRS

Eshel="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Eshel/scDRS/Eshel_et_al.h5ad" # celltype
Milon="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Milon/scDRS/Milon_et_al.h5ad" # celltype
Ranum="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Ranum/scDRS/Ranum_et_al.h5ad" # cell_type
Jean="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Jean/scDRS/Jean_et_al_SYMBL.h5ad" # cell_type


Sun1_2="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Sun/1-2M/scDRS/age_1M_2M.h5ad" # CellType
Sun12_15="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Sun/12-15M/scDRS/age_12M_15M.h5ad" # CellType
Sun5="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Sun/5M/scDRS/age_5M.h5ad" # CellType
Xu="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Xu/scDRS/Xu_et_al.h5ad" # celltype
Iyer="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/Iyer/scDRS/Iyer_et_al.h5ad" # cell_type


expr=$Jean
geneset="/public/home/shilulu/Wulab_project/ARHL/NC_sup_test/08.scEnrich/config/ARHL.scdrs"
qsubshcom "scdrs compute-score \
    --h5ad-file $expr \
    --h5ad-species mouse \
    --gs-file $geneset \
    --gs-species human \
    --flag-filter-data True \
    --flag-raw-count False \
    --flag-return-ctrl-raw-score False \
    --flag-return-ctrl-norm-score True \
    --out-folder ./" 1 100G scdrs 5:00:00 ""

# for cell type level
expr=$Jean
qsubshcom "scdrs perform-downstream \
    --h5ad-file $expr \
    --score-file ./ARHL.full_score.gz \
    --group-analysis cell_type \
    --flag-raw-count False \
    --out-folder ./" 1 100G scdrs 5:00:00 ""


# 4. plot scDRS
dict_score = {
    trait: pd.read_csv(f"single_cell_data/cwy/{trait}.full_score.gz", sep="\t", index_col=0)
    for trait in df_gs.index}

for trait in dict_score:
    adata.obs[trait] = dict_score[trait]["norm_score"]

sc.set_figure_params(figsize=[2.5, 2.5], dpi=150)
sc.pl.umap(
    adata,
    color="level1class",
    ncols=1,
    color_map="RdBu_r",
    vmin=-5,
    vmax=5)

sc.pl.umap(
    adata,
    color=dict_score.keys(),
    color_map="RdBu_r",
    vmin=-5,
    vmax=5,
    s=20)
